synapse/handlers/event_auth.py

```diff
@@ -29,44 +31,59 @@ class EventAuthHandler:
     def __init__(self, hs: "HomeServer"):
         self._store = hs.get_datastore()
 
-    async def can_join_without_invite(
-        self, state_ids: StateMap[str], room_version: RoomVersion, user_id: str
-    ) -> bool:
+    async def check_restricted_join_rules(
+        self,
+        state_ids: StateMap[str],
+        room_version: RoomVersion,
+        user_id: str,
+        prev_member_event: Optional[EventBase],
+    ) -> None:
         """
-        Check whether a user can join a room without an invite.
+        Check whether a user can join a room without an invite due to restricted join rules.
 
         When joining a room with restricted joined rules (as defined in MSC3083),
-        the membership of spaces must be checked during join.
+        the membership of spaces must be checked during a room join.
 
         Args:
             state_ids: The state of the room as it currently is.
             room_version: The room version of the room being joined.
             user_id: The user joining the room.
+            prev_member_event: The current membership event for this user.
 
-        Returns:
-            True if the user can join the room, false otherwise.
+        Raises:
+            AuthError if the user is cannot join the room.
         """
+        # If the member is invited or currently joined, then nothing to do.
+        if prev_member_event and (
+            prev_member_event.membership == Membership.JOIN
+            or prev_member_event.membership == Membership.INVITE
```

Comment:
maybe?

```suggestion
            prev_member_event.membership in (Membership.JOIN, Membership.INVITE)
```

Reply:
I purposefully didn't do this, but don't have a strong opinion on it. I'd really love to do it as a set comparison but that causes issues if `membership` is garbage data (and not hashable). 🤷

Reply:
Why would set comparison be better here? (I'm actually surprised to discover that a set comparison here *is* faster, given traditionally for small numbers of values lists are faster than sets :shrug: )

Reply:
#7737 has some data.

Reply:
(I didn't test which would be faster by the way! Just that we seem to be fairly inconsistent with how we do this...I also feel like we should pull out constant literals like this to avoid re-allocating them frequently...but maybe that's an over-optimization.)